AppController/djinn.rb

```diff
@@ -1137,38 +1137,44 @@ def get_cluster_stats_json(secret)
       end
     end
 
-    JSON.dump(@cluster_stats)
+    @state_change_lock.synchronize { return JSON.dump(@cluster_stats) }
   end
 
   # Updates our locally cached information about the CPU, memory, and disk
   # usage of each machine in this AppScale deployment.
   def update_node_info_cache
+    threads = []
     new_stats = []
-
-    Thread.new {
+    @state_change_lock.synchronize {
```

Comment:
Why do these fetches need to be under the state change lock?

Reply:
we are looping through @nodes, which is  a shared variable and could change (say add_nodes etc...)

Reply:
Probably: since we do this just about anywhere else in the djinn code, I was thinking to leave that to a different PR that works around all the Mutexes not just this one. I'm happy to get it started here though it you think it makes sense.

Reply:
The reason I bring it up for this PR is that you wrapped several other places with the state change lock. For example, `appscale status` now has the potential to hang for 10 more seconds than it did before.

I tried looking into whether `JSON.dump` was atomic (locking the hash while iterating it, for example), but I couldn't find any readily available answers.

Reply:
Got it. Let me add something now.

Reply:
@cdonati let me know if something like this makes sense to you or you prefer something simpler. I ended up spawning a thread to update nodes, to ensure we don't lock the duty cycle no matter what (say badly behavior network) and limit the holding of the lock within the update method.